### ISSUE TITLE
Confusion in equation 2

### ISSUE THREAD

**Issue by Gouiaa** [xingyizhou/CenterNet#37]
@xingyizhou Thanks for sharing the code and the paper.
just I want to check the equation 2. The matrix `O=(P/R -p_tilde)` with dimension `W/RxH/Rx2` contains the offset correspond to each center in the image. Example: if we have an image (512, 512) which contains 1 object whose center is `p = (130, 130)`. So `p_tilde=floor((130/4, 130/4)) =(32, 32)` so `p/R -p_tilde =(0.5, 0.5)`. So the position (32, 32, 0) and (32, 32, 1) (index start with 0 like in python) in the matrix`O ` contains (.5, 0.5) and the rest of positions are set to 0. Is it right? 

Another question: why all centers share the same tensor? I am not sure but maybe it will be difficult to map the offset back to the centers, particularly when we have two or more centers that are very close to each others !? 
Another small question: in equation 1 what is the subscript `k` in` L_k` ?

**Comment by xingyizhou** [comment]
Hi,
  Thanks for the detailed reading.
  1. Yes. And other positions will be masked out.
  2. We follow CornerNet.Actually such collisions happen very rarely, even in the category-agnostic case. You can refer to Appendix C of our paper and check the collision experiments code [here](https://github.com/xingyizhou/CenterNet/blob/master/src/tools/calc_coco_overlap.py#L117).
  3. `k` is short for `keypoints`, similar to `L_{size}` in eq. 2.

**Comment by Gouiaa** [comment]
Thanks @xingyizhou 
I have an other question in the section "From points to bounding boxes" about how to produce the  bounding box (see the equation at the bottom of  `page 3 just after equation 4`. I think you should annotate this equation as equation 5  :)  ).
 `x_hat_i , delta_x_hat_i` are  inferred in the dimension of the heat map `(W/R, H/R)`, however `w_hat_i `and `h_hat_i `are inferred in the original size (see `L_size` in `equation 3`). So in the `equation 5`, `w_hat_i `and `h_hat_i` should be divided by `R`. What do you think ?

**Comment by xingyizhou** [comment]
Hi,
  Thanks for the suggestion. Yes, "formula 5" is in the output size. The final prediction is multiplied by `R`.

**Comment by Gouiaa** [comment]
Okay, you mean (x_hat_i+ delta_x_hat_i) and (y_hat_i+ delta_y_hat_i) are multiplied by R ?

**Comment by xingyizhou** [comment]
I mean we will multiply the whole entity by `R` after "formula 5".

**Comment by Gouiaa** [comment]
Sorry but I think it  does not make sense. You have to multiply only the (x_hat_i+ delta_x_hat_i) and (y_hat_i+ delta_y_hat_i) by R, because w_hat and h_hat are already in the original size. Thanks

**Comment by xingyizhou** [comment]
Oh actually in our [implementation](https://github.com/xingyizhou/CenterNet/blob/master/src/lib/datasets/sample/ctdet.py#L105), `w` and `h` are in the downsampled space. Sorry that it is not clearly stated in the paper. We will add this in the next revision.

**Comment by Gouiaa** [comment]
Thanks for you time @xingyizhou . I am reading the part of the 3D estimation. Your estimation is based on the paper in [13]. I read your supplementary which is very helpfull and I understand that you estimate the depth that correspond to every center point, also the dimension and the angle.

My questions (or remarks ) are:
1) I don't understand  `d=  1/σ(ˆd)−1` and how this quantity is in [0, 1] (this quantity is in [-2, 0]).

> We instead use the output transformation  of  Eigenet  al.  [13]  and d=  1/σ(ˆd)−1,  where σ is the sigmoid function.  We compute the depth as an additional  output  channel ˆD ∈ [0,1] WR×HRof  our  keypointestimator.

2) I checked the paper [13] and If I understand well they used a sort of regression with a particular loss function that takes into account the relation between pairs of pixels in ground truth and in  the predicted depth map ( each pair of pixels in the prediction must differ in depth by an amount similar to that of the corresponding pair in the ground truth). I don't see exactly where you used the paper [13] depth estimation ?
3) The orientation angle is with respect to the camera or  with respect to the ray that goes through the center of the bounding-box ?

**Comment by xingyizhou** [comment]
1. Sorry, Depth should be in R ^ WR×HR. We have fixed this in the supplementary but missed the main paper. We will fix this in the next revision.
2. We mean we take the idea of estimating the inverse depth from [13]. 
3. The network predicts the camera-view vehicle orientation, and the orientation is then converted to the global orientation [here](https://github.com/xingyizhou/CenterNet/blob/master/src/lib/utils/ddd_utils.py#L77). You can check Fig.3, Fig.4 of [Mou. et al](https://arxiv.org/pdf/1612.00496.pdf) and Sec. 4.2 (Allocentric vs. Egocentric:) of [3DRCNN](http://openaccess.thecvf.com/content_cvpr_2018/papers/Kundu_3D-RCNN_Instance-Level_3D_CVPR_2018_paper.pdf) for more details about the orientation representation.
4. You can run `python test.py ddd --dataset kitti --load_model ../models/ddd_3dop.pth --debug 2` to visualize the 3D bbox output on KITTI, after setting up kitti data. I haven't tried our 3D model on custom images. I don't expect it will work well on images outside kitti because kitti is in a strange aspect ratio and requires a known camera intrinsic.

**Comment by Gouiaa** [comment]
@xingyizhou This paper seems very good but it needs to be re-written from scratch!

**Comment by xingyizhou** [comment]
Hi,
  Thanks for your feedback. We will more appreciate it if you can specify which aspects of the writing you don't like.

**Comment by Gouiaa** [comment]
@xingyizhou, I think you have to fix the points that I have mentioned and other things is you can add more understandable figures for the models (because you did not handle the model visualization in the code). Thanks

**Comment by ustczhouyu** [comment]
Training and testing on my own data, many objects are missed, but when each image is cut into two images and then tested, there is no missed objects. Maybe it is because my data contains too many objects. How to modify the object number of output ?
